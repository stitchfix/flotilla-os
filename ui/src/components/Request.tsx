import * as React from "react"
import { AxiosError } from "axios"

export enum RequestStatus {
  READY = "READY",
  NOT_READY = "NOT_READY",
  ERROR = "ERROR",
}

export type Props<ResponseType, ArgsType> = {
  children: (props: ChildProps<ResponseType, ArgsType>) => React.ReactNode
  requestFn: (args: ArgsType) => Promise<ResponseType>
  initialRequestArgs: ArgsType
  shouldRequestOnMount: boolean
  onSuccess?: (res: ResponseType) => void
  onFailure?: (error: any) => void
}

export type State<ResponseType> = {
  requestStatus: RequestStatus
  data: ResponseType | null
  isLoading: boolean
  error: AxiosError | null
}

export type ChildProps<ResponseType, ArgsType> = State<ResponseType> & {
  request: (opts: ArgsType) => void
}

class Request<ResponseType, ArgsType> extends React.Component<
  Props<ResponseType, ArgsType>,
  State<ResponseType>
> {
  static defaultProps = {
    shouldRequestOnMount: true,
    initialRequestArgs: null,
  }

  state = {
    requestStatus: RequestStatus.NOT_READY,
    data: null,
    isLoading: false,
    error: null,
  }

  componentDidMount() {
    if (this.props.shouldRequestOnMount) {
      this.request(this.props.initialRequestArgs)
    }
  }

  request(args: ArgsType): void {
    const { requestFn, onSuccess, onFailure } = this.props

    this.setState({ isLoading: true })

    requestFn(args)
      .then((data: ResponseType) => {
        this.setState({
          data,
          isLoading: false,
          requestStatus: RequestStatus.READY,
          error: null,
        })
        if (onSuccess) onSuccess(data)
      })
      .catch((error: AxiosError) => {
        this.setState({
          isLoading: false,
          requestStatus: RequestStatus.ERROR,
          error,
        })
        if (onFailure) onFailure(error)
      })
  }

  getChildProps = () => ({
    ...this.state,
    request: this.request.bind(this),
  })

  render() {
    return this.props.children(this.getChildProps())
  }
}

export default Request
