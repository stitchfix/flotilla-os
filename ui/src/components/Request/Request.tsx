import * as React from "react"
import { isArray } from "lodash"
import { flotillaUIRequestStates } from "../../types"

type RequestArgs = { [key: string]: any }
type RequestFn = (args?: any) => Promise<any>

export interface IProps {
  shouldRequestOnMount: boolean
  initialRequestArgs?: RequestArgs | RequestArgs[]
  requestFn: RequestFn | RequestFn[]
  children: (props: IChildProps) => React.ReactNode
}

export interface IState {
  inFlight: boolean
  data: any
  requestState: flotillaUIRequestStates
  error: any
}

export interface IChildProps extends IState {
  request: (args?: any) => void
}

class Request extends React.PureComponent<IProps, IState> {
  static defaultProps: Partial<IProps> = {
    shouldRequestOnMount: true,
  }

  state = {
    inFlight: false,
    data: null,
    requestState: flotillaUIRequestStates.NOT_READY,
    error: false,
  }

  componentDidMount() {
    const { shouldRequestOnMount, initialRequestArgs } = this.props

    if (shouldRequestOnMount) {
      this.request(initialRequestArgs)
    }
  }

  request(requestArgs?: RequestArgs | RequestArgs[]): void {
    const { requestFn } = this.props
    this.setState({ inFlight: true })

    if (isArray(requestFn) && isArray(requestArgs)) {
      Promise.all(requestFn.map((fn, i) => fn(requestArgs[i])))
        .then(data => {
          this.setState({
            data,
            inFlight: false,
            error: false,
            requestState: flotillaUIRequestStates.READY,
          })
        })
        .catch(error => {
          this.setState({
            inFlight: false,
            error,
            requestState: flotillaUIRequestStates.ERROR,
          })
        })
    } else if (!isArray(requestFn) && !isArray(requestArgs)) {
      requestFn(requestArgs)
        .then(data => {
          this.setState({
            data,
            inFlight: false,
            error: false,
            requestState: flotillaUIRequestStates.READY,
          })
        })
        .catch(error => {
          this.setState({
            inFlight: false,
            error,
            requestState: flotillaUIRequestStates.ERROR,
          })
        })
    } else {
      console.error(
        `The requestFn and requestArgs props passed to <Request> must either both be arrays or non arrays.`
      )
    }
  }

  render() {
    return this.props.children({
      ...this.state,
      request: this.request,
    })
  }
}

export default Request
