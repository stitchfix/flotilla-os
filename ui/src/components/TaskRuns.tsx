import * as React from "react"
import { Link } from "react-router-dom"
import { get, omit } from "lodash"
import ListRequest, { ChildProps as ListRequestChildProps } from "./ListRequest"
import api from "../api"
import {
  ListTaskRunsParams,
  ListTaskRunsResponse,
  SortOrder,
  Run,
} from "../types"
import pageToOffsetLimit from "../helpers/pageToOffsetLimit"
import Table from "./Table"
import { FormGroup, Classes } from "@blueprintjs/core"
import GenericMultiSelect from "./GenericMultiSelect"
import RunStatusSelect from "./RunStatusSelect"
import ListFiltersDropdown from "./ListFiltersDropdown"
import { DebounceInput } from "react-debounce-input"
import Pagination from "./Pagination"

export const TaskRuns: React.FunctionComponent<
  ListRequestChildProps<ListTaskRunsResponse, { params: ListTaskRunsParams }>
> = ({
  data,
  updateSort,
  currentSortKey,
  currentSortOrder,
  query,
  updateFilter,
  updatePage,
  currentPage,
}) => (
  <>
    <div className="flotilla-list-utils-container">
      <FormGroup label="Run Status" helperText="Search by run status.">
        <RunStatusSelect
          value={get(query, "status", [])}
          onChange={(value: string[]) => {
            updateFilter("status", value)
          }}
        />
      </FormGroup>
      <ListFiltersDropdown>
        <FormGroup label="Cluster" helperText="Search by ECS cluster.">
          <GenericMultiSelect
            value={get(query, "cluster", [])}
            onChange={(value: string[]) => {
              updateFilter("cluster", value)
            }}
          />
        </FormGroup>
        <FormGroup
          label="Started At Since"
          helperText="Enter a valid ISO8601 string."
        >
          <DebounceInput
            style={{ flex: 1 }}
            className={Classes.INPUT}
            debounceTimeout={500}
            value={get(query, "started_at_since", "")}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => {
              updateFilter("started_at_since", evt.target.value)
            }}
          />
        </FormGroup>
        <FormGroup
          label="Started At Until"
          helperText="Enter a valid ISO8601 string."
        >
          <DebounceInput
            style={{ flex: 1 }}
            className={Classes.INPUT}
            debounceTimeout={500}
            value={get(query, "started_at_until", "")}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => {
              updateFilter("started_at_until", evt.target.value)
            }}
          />
        </FormGroup>
        <FormGroup
          label="Finished At Since"
          helperText="Enter a valid ISO8601 string."
        >
          <DebounceInput
            style={{ flex: 1 }}
            className={Classes.INPUT}
            debounceTimeout={500}
            value={get(query, "finished_at_since", "")}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => {
              updateFilter("finished_at_since", evt.target.value)
            }}
          />
        </FormGroup>
        <FormGroup
          label="Finished At Until"
          helperText="Enter a valid ISO8601 string."
        >
          <DebounceInput
            style={{ flex: 1 }}
            className={Classes.INPUT}
            debounceTimeout={500}
            value={get(query, "finished_at_until", "")}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => {
              updateFilter("finished_at_until", evt.target.value)
            }}
          />
        </FormGroup>
      </ListFiltersDropdown>
      <Pagination
        updatePage={updatePage}
        currentPage={currentPage}
        totalPages={data ? data.total : 1}
      />
    </div>
    <Table<Run>
      items={get(data, "history", [])}
      getItemKey={(r: Run) => r.run_id}
      updateSort={updateSort}
      currentSortKey={currentSortKey}
      currentSortOrder={currentSortOrder}
      columns={{
        run_id: {
          displayName: "Run ID",
          render: (r: Run) => <Link to={`/runs/${r.run_id}`}>{r.run_id}</Link>,
          isSortable: true,
        },
        status: {
          displayName: "Status",
          render: (r: Run) => r.status,
          isSortable: true,
        },
        started_at: {
          displayName: "Started At",
          render: (r: Run) => r.started_at || "-",
          isSortable: true,
        },
        finished_at: {
          displayName: "Finished At",
          render: (r: Run) => r.finished_at || "-",
          isSortable: true,
        },
        cluster: {
          displayName: "Cluster",
          render: (r: Run) => r.cluster,
          isSortable: false,
        },
      }}
    />
  </>
)

const ConnectedTaskRuns: React.FunctionComponent<{ definitionID: string }> = ({
  definitionID,
}) => (
  <ListRequest<
    ListTaskRunsResponse,
    { definitionID: string; params: ListTaskRunsParams }
  >
    requestFn={api.listTaskRuns}
    initialQuery={{
      page: 1,
      sort_by: "started_at",
      order: SortOrder.DESC,
    }}
    getRequestArgs={params => ({
      definitionID,
      params: {
        ...omit(params, "page"),
        ...pageToOffsetLimit({ page: get(params, "page", 1), limit: 20 }),
      },
    })}
  >
    {props => <TaskRuns {...props} />}
  </ListRequest>
)

export default ConnectedTaskRuns
