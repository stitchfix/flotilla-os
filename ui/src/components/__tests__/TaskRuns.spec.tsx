import React from "react"
import { mount } from "enzyme"
import { mountToJson } from "enzyme-to-json"
import { TaskRuns } from "../TaskRuns"
import { RequestStatus } from "../Request"

describe("TaskRuns", () => {
  it("renders", () => {
    const wrapper = mount(
      <TaskRuns
        requestStatus={RequestStatus.NOT_READY}
        data={null}
        isLoading={false}
        error={false}
        updateSort={jest.fn()}
        updatePage={jest.fn()}
        updateFilter={jest.fn()}
      />
    )
    expect(mountToJson(wrapper)).toMatchSnapshot()
  })
})
