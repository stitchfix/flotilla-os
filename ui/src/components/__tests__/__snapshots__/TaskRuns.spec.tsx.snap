// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskRuns renders 1`] = `
<TaskRuns
  data={null}
  error={false}
  isLoading={false}
  requestStatus="NOT_READY"
  updateFilter={[MockFunction]}
  updatePage={[MockFunction]}
  updateSort={[MockFunction]}
>
  <Table
    columns={
      Object {
        "cluster": Object {
          "displayName": "Cluster",
          "isSortable": false,
          "render": [Function],
        },
        "finished_at": Object {
          "displayName": "Finished At",
          "isSortable": true,
          "render": [Function],
        },
        "run_id": Object {
          "displayName": "Run ID",
          "isSortable": true,
          "render": [Function],
        },
        "started_at": Object {
          "displayName": "Started At",
          "isSortable": true,
          "render": [Function],
        },
        "status": Object {
          "displayName": "Status",
          "isSortable": true,
          "render": [Function],
        },
      }
    }
    getItemKey={[Function]}
    items={Array []}
    updateSort={[MockFunction]}
  >
    <HTMLTable
      bordered={true}
      striped={true}
    >
      <table
        className="bp3-html-table bp3-html-table-bordered bp3-html-table-striped"
      >
        <thead>
          <tr>
            <Th
              isActive={false}
              isSortable={true}
              key="run_id"
              onClick={[Function]}
            >
              <th
                className="flotilla-th-sortable"
                onClick={[Function]}
              >
                Run ID
              </th>
            </Th>
            <Th
              isActive={false}
              isSortable={true}
              key="status"
              onClick={[Function]}
            >
              <th
                className="flotilla-th-sortable"
                onClick={[Function]}
              >
                Status
              </th>
            </Th>
            <Th
              isActive={false}
              isSortable={true}
              key="started_at"
              onClick={[Function]}
            >
              <th
                className="flotilla-th-sortable"
                onClick={[Function]}
              >
                Started At
              </th>
            </Th>
            <Th
              isActive={false}
              isSortable={true}
              key="finished_at"
              onClick={[Function]}
            >
              <th
                className="flotilla-th-sortable"
                onClick={[Function]}
              >
                Finished At
              </th>
            </Th>
            <Th
              isActive={false}
              isSortable={false}
              key="cluster"
              onClick={[Function]}
            >
              <th
                className=""
                onClick={[Function]}
              >
                Cluster
              </th>
            </Th>
          </tr>
        </thead>
        <tbody />
      </table>
    </HTMLTable>
  </Table>
</TaskRuns>
`;
